### ISSUE TITLE
key mismatch on network_interface.0 (old: 0, new: 4000). This is a bug with the provider, please report it

### ISSUE THREAD

**Issue by sofixa** [terraform-providers/terraform-provider-vsphere#654]
### Terraform Version
0.11.10

### vSphere Provider Version

```
- Downloading plugin for provider "vsphere" (1.9.0)...
```

### Affected Resource(s)

- `vsphere_virtual_machine`

### Terraform Configuration Files

```hcl
resource "vsphere_virtual_machine" "pa2-elk" {
  count            = "${var.pa2-elk_count}"
  provider         = "vsphere.pa2"
  name             = "${var.client}.${var.env}.${var.name_dataroom}-${var.elk_project1_instance}${count.index + 1 + var.pa2-elk_count-offset}"
  guest_id         = "${data.vsphere_virtual_machine.pa2_template9.guest_id}"
  resource_pool_id = "${data.vsphere_resource_pool.pa2_pool.id}"
  datastore_id     = "${data.vsphere_datastore.pa2_datastore-01.id}"
  folder           = "${var.client}"

  num_cpus = "${var.elk_num-cpus}"
  memory   = "${var.elk_memory * 1024}"

  network_interface {
    network_id   = "${data.vsphere_network.pa2_adm.id}"
    adapter_type = "${var.network_interface_type}"
  }

  network_interface {
    network_id   = "${data.vsphere_network.pa2_project1_back.id}"
    adapter_type = "${var.network_interface_type}"
  }

  network_interface {
    network_id   = "${data.vsphere_network.pa2_adm.id}"
    adapter_type = "${var.network_interface_type}"
  }


  disk {
    label            = "disk0.vmdk"
    size             = "${var.elk_disk-size}"
    eagerly_scrub    = "${var.eagerly_scrub}"
    thin_provisioned = "${var.thin_provisioned}"
  }

  lifecycle {
    ignore_changes = [ "clone.0.template_uuid","disk.0" ]
  }

  cpu_hot_add_enabled        = true
  memory_hot_add_enabled     = true
  wait_for_guest_net_timeout = 0

  clone {
    template_uuid = "${data.vsphere_virtual_machine.pa2_template9.id}"
  }

  extra_config {
    "guestinfo.vmname" = "${var.client}.${var.env}.${var.name_dataroom}-${var.elk_project1_instance}${count.index + 1 + var.pa2-elk_count-offset}"
  }
}
```
Nothing special.

### Debug Output

https://gist.github.com/sofixa/db36a3ab462fd94be67538c54de817b5

### Expected Behavior

When adding a new NIC or modifying an existing one, well, i expected the addition/modification will happen.

### Actual Behavior

```
key mismatch on network_interface.0 (old: 0, new: 4000). This is a bug with the provider, please report it
```

### Steps to Reproduce

1. Create a VM
2. Add/modify a NIC (not sure when exactly does it happen - we've had it happen 3 times with nothing in common, on things updated from older provider version and relatively new ones built on 1.6, then had the bug happen, then updated to 1.9 and the bug is still there)
3. `terraform apply`

What's weird is that in the tfstate the key is already the "new one" - 4000. Furthermore, if there are no modifications on the NICs, a plan/apply passes without any issues.
 
### References

The error literally says "this is a bug with the provider, please report it" :D 

So, what can we do to:

* fix this in a non-disruptive manner (IIRC we fixed it the first time by deleting all NICs in vSphere, deleting them from the tfstate and re-plan/applying, but that's not an actual solution)
* make sure it doesn't happen again?

Cheers,
Adrian

**Comment by sofixa** [comment]
Anyone? This seems to be a known bug because the code literally states "This is a bug with the provider, please report it". Is there some sort of workaround, solution for this?

ping @bill-rich ?

**Comment by bill-rich** [comment]
@sofixa So far, I've been unable to reproduce the issue. I'm not giving up on it though. I am guessing that the list of nics is getting out of order during while matching up which are new and which already exist. The trick I believe is figuring out what exactly is causing this in only very limited cases.

When a VM exhibits this issue, you consistently will get this error when trying to apply?

**Comment by sofixa** [comment]
@bill-rich , yes, it's consistent and persist through deleting all NICs in terraform configuration or terraform state; if we remove the new NICs it happily plans/applies, so it's the addition that seems to mess things up. The only thing that seems to fix it is deleting all NICs from VMware ( and it's a temporary fix, re-adding a new NIC messes things up again). It's happened 4 times for a few months and a few hundred VMs, but not a lot of NIC modifications.

If you want i can provide the output of a govc vm.info (detailed one) of a "broken" VM and a working one to compare if there's something weird/different NIC-wise ?

**Comment by congiohj** [comment]
Hello @bill-rich,

I'm a coworker of @sofixa 
[Here](https://gist.github.com/congiohj/0204aed90445d301df77da006c3b78ba) is a gist with the govc.info output for both vm.

The first output is for the broken vm, the seconde is for the working one

**Comment by bill-rich** [comment]
Hi @sofixa and @congiohj! 

Sorry for the delay in getting back to you on this. I've been revisiting the issue and have still been unable to reproduce it, but have improved diff logging in v1.9.1+ of the provider. If you still have access to VMs exhibiting the issue, can you please upgrade to the latest vSphere provider and upload a new debug log for me? Hopefully with the improved logging, we'll be able to figure out exactly what is going on.

**Comment by congiohj** [comment]
Hi @bill-rich 

I have upgraded the provider to 1.10, but we still have the issue.

 [Here](https://gist.github.com/congiohj/8ef33a2d50f2610565ef6fc5c9bb8c56) is a gist with a terraform plan.

**Comment by DjinnS** [comment]
Hi,

I work with @sofixa and we did some tests and find that the issue seems related with the template used by the VM.

If the template_uuid for the VM (in tfstate) is the same that the template_uuid in the data source, then, there is no problem when applying the new configuration.
If a VM (already existing, so, present in the tfstate) doesn't have the same template_uuid than the data source, then, then apply will failed as described in the issue.

So we have a workaround by manually changing the template_uuid, but we don't understand this behaviour. Is someone have an explanation ? Is there an another workaround than change a value in the tfstate ? We have to do this action every time we push a new template ...

**Comment by sofixa** [comment]
Hi @bill-rich , there is also #357 and #527 which seem similar.

After a high-level look at the DEBUG output of terraform, it seems that on every plan after the diff customization and validation it also does a `DiskCloneValidateOperation` and `ValidateVirtualMachineClone`, even if that isn't needed, and it talks about new resource sets (`DiskDiffOperation: New resource set post-normalization: (key 2000 at scsi:0:0)`)

To reproduce, you just need to create a VM from a template, and change the template's UUID/rebuild it, which will result in a bunch of weirdness during the plan regarding the disks (new keys), even if nothing has changed in the template, and a failure on apply with overlapping/broken keys/ids on disks and networking interfaces if you add a new one.

Considering we rebuild our templates regularly via Packer, this makes the provider basically unusable for us. Could you take a look

**Comment by rismoney** [comment]
@DjinnS : relating to the template uuid - https://github.com/terraform-providers/terraform-provider-vsphere/issues/353

tldr: use uuid specification directly instead of data sources against template names.

**Comment by sofixa** [comment]
@rismoney that really doesn't scale well though, compared to a data source, which would always be the latest version for new builds.

**Comment by rismoney** [comment]
1. if the template changes - do you want a rebuild or no?  If the answer if yes, do nothing.  If it's no

A. I want a failure that tells me the template no longer exists.
B. I want to be deterministic and specify the template by what it is (uuid specify)
C. Ignore template changes altogether is the other alternative.

Personally we have moved to direct uuid specification.  It does force us to always know which template we're deploying by guid.   But as we manage everything in git, we become very explicit about changes like template updates and accountability.  So when we build templates, we put annotations for the uuid so they are easily viewable.  Then we treat them like the atomic units we believe they are.

In thinking about this, perhaps https://www.terraform.io/docs/providers/terraform/d/remote_state.html could help in getting the value data you're looking for through state with a conditional.  Not sure - just throwing it out there.

**Comment by sofixa** [comment]
@rismoney we're in case C. Our workflow is the following:

* Packer builds templates regularly and a script pushes them to ~30 different vCenters
* terraform uses a data source to get the latest template UUID to use it on new VM creations (with a lifecycle ignore_change on template_uuid to not re-create VMs if the template UUID changes unless needed)

Using fixed UUID manually would break our modules (since they're used on different vCenters via a dynamic data source and template UUIDs aren't the same across vCenters) on top of adding a manual stage.

If we just put the latest UUID in Vault or somewhere else accessible via a data source (like remote state you proposed), we're back to square one - when the UUID that the data source returns changes, the VMs will again be broken.

In any case, @bill-rich , IMHO the current behaviour is broken - considering the purpose of the lifecycle ignore_change block, it's not normal that things are modified (to be precise, terraform tries to modify them and fails) based on information (a dynamic data source) that should be ignored.

**Comment by bill-rich** [comment]
Same root cause as #357. Closing to track the issue there.
